executor/cook/executor.py

```diff
@@ -0,0 +1,364 @@
+
+
+class ProgressWatcher(object):
+    """
+        This class tails the output from the target file listening for progress messages.
+        The retrieve_progress_states generates all progress messages iteratively.
+    """
+    @staticmethod
+    def match_progress_update(input_string):
+        result = re.findall(r'\^\^\^\^JOB-PROGRESS: (\d*)(?: )?(.*)', input_string)
+        return result[0] if len(result) >= 1 else None
+
+    def __init__(self, target_file, completed_signal):
+        self.target_file = target_file
+        self.progress = None
+        self.completed_signal = completed_signal
+
+    def current_progress(self):
+        return self.progress
+
+    def tail(self, sleep_time_ms):
+        while True:
+            if os.path.isfile(self.target_file):
+                logging.info('{} has been created, reading contents'.format(self.target_file))
+                break
+            if self.completed_signal.isSet():
+                logging.info('No output has been read to parse progress messages')
+                return
+            logging.info('{} has not yet been created, sleeping {} ms'.format(self.target_file, sleep_time_ms))
+            time.sleep(sleep_time_ms / 1000.0)  # sleep briefly
+
+        target_file_obj = open(self.target_file, 'r')
+        while True:
+            line = target_file_obj.readline()
+            if not line:
+                if self.completed_signal.isSet():
+                    target_file_obj.close()
+                    break
+                time.sleep(sleep_time_ms / 1000.0)  # sleep briefly
+                continue
+            yield line
+
+    def retrieve_progress_states(self):
+        for line in self.tail(50):
+            progress_report = ProgressWatcher.match_progress_update(line)
+            if progress_report is not None:
+                percent, message = progress_report
+                logging.info('Updating progress to {} percent, message: {}'.format(percent, message))
+                self.progress = {'progress-message': message.strip(), 'progress-percent': int(percent)}
+                yield self.progress
+
+
+class ProgressUpdater(object):
+    """
+        This class is responsible for sending progress updates to the scheduler.
+        It throttles the rate at which progress updates are sent.
+    """
+    def __init__(self, poll_interval_ms):
+        self.poll_interval_ms = poll_interval_ms
+        self.last_reported_time = None
+        self.last_progress_data = None
+        self.lock = Lock()
+
+    def has_enough_time_elapsed_since_last_update(self):
+        if self.last_reported_time is None:
+            return True
+        else:
+            current_time = time.time()
+            time_diff_ms = (current_time - self.last_reported_time) * 1000
+            return time_diff_ms >= self.poll_interval_ms
+
+    def send_progress_update(self, driver, task_id, max_message_length, progress_data, force_send=False):
+        with self.lock:
```

Comment:
`send_progress_update` can potentially be called from two threads - the one doing the progress update, the one handling termination and force sending the current progress state.